fedimint-client/src/lib.rs

```diff
@@ -317,20 +323,88 @@ pub type ModuleGlobalContextGen = ContextGen<DynGlobalClientContext>;
 impl Client {
     /// Add funding and/or change to the transaction builder as needed, finalize
     /// the transaction and submit it to the federation.
-    pub async fn finalize_and_submit_transaction(
+    pub async fn finalize_and_submit_transaction<F, M>(
         &self,
         operation_id: OperationId,
+        operation_type: &str,
+        operation_meta: F,
         tx_builder: TransactionBuilder,
-    ) -> anyhow::Result<TransactionId> {
+    ) -> anyhow::Result<TransactionId>
+    where
+        F: Fn(TransactionId) -> M,
+        M: serde::Serialize,
+    {
         let mut dbtx = self.inner.db.begin_transaction().await;
+
+        if ClientInner::operation_exists(&mut dbtx, operation_id).await {
+            bail!("There already exists an operation with id {operation_id:?}")
+        }
+
         let txid = self
             .inner
             .finalize_and_submit_transaction(&mut dbtx, operation_id, tx_builder)
             .await?;
+
+        dbtx.insert_entry(
+            &OperationLogKey { operation_id },
+            &OperationLogEntry {
+                operation_type: operation_type.to_string(),
+                meta: serde_json::to_value(operation_meta(txid))?,
+            },
+        )
+        .await;
+        dbtx.insert_entry(
+            &ChronologicalOperationLogKey {
+                creation_time: fedimint_core::time::now(),
+                operation_id,
+            },
+            &(),
+        )
+        .await;
+
         dbtx.commit_tx().await;
         Ok(txid)
     }
 
+    // TODO: allow fetching pages
```

Comment:
Note: don't use pages, use cursors. I thought it might be worth mentioning...

BTW. instead of pages&cursor ... just return the iterator directly to the caller maybe (with all the `.map` etc needed)?

Reply:
I tried, but for that to be possible we need read-only database transactions such that we can read from the prefix stream while also querying the same dbtx for operation details. Currently that's impossible because both the stream and the read fns take mut refs. That's also necessary for a lot of use cases because database transactions aren't sync so non-mut refs are not send. So maybe it wouldn't even work with a read-only database transaction?

Reply:
@elsirion  I was thinking that we could return a tuple of database transaction and the iterator. I think it should work, where the lifetime of the iterator would be marked as shorter than of the dbtx.